src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/AssetsFileDependenciesTreeSearchProvider.cs

```diff
@@ -0,0 +1,191 @@
+
+        public async Task SearchAsync(IDependenciesTreeProjectSearchContext context)
+        {
+            // get latest snapshot
+            ExportProvider exportProvider = context.UnconfiguredProject.Services.ExportProvider;
+
+            Lazy<IAssetsFileDependenciesDataSource, IAppliesToMetadataView> dataSource
+                = exportProvider
+                    .GetExports<IAssetsFileDependenciesDataSource, IAppliesToMetadataView>()
+                    .SingleOrDefault(export => export.Metadata.AppliesTo(context.UnconfiguredProject.Capabilities));
+
+            if (dataSource == null)
+            {
+                // dataSource will be null for shared projects, for example
+                return;
+            }
+
+            IProjectDataSourceRegistry? dataSourceRegistry = context.UnconfiguredProject.Services.DataSourceRegistry;
+            Assumes.Present(dataSourceRegistry);
+
+            AssetsFileDependenciesSnapshot snapshot = (await dataSource.Value.GetLatestVersionAsync<AssetsFileDependenciesSnapshot>(dataSourceRegistry, cancellationToken: context.CancellationToken)).Value;
+
+            if (!(context.UnconfiguredProject.Services.ExportProvider.GetExportedValue<IActiveConfigurationGroupService>() is IActiveConfigurationGroupService3 activeConfigurationGroupService))
+            {
+                return;
+            }
+
+            IConfigurationGroup<ConfiguredProject> configuredProjects = await activeConfigurationGroupService.GetActiveLoadedConfiguredProjectGroupAsync();
+
+            foreach ((_, AssetsFileTarget target) in snapshot.DataByTarget)
+            {
+                ConfiguredProject? configuredProject = await FindConfiguredProjectAsync(target.TargetFrameworkMoniker);
+
+                if (configuredProject == null)
+                {
+                    continue;
+                }
+
+                IDependenciesTreeConfiguredProjectSearchContext? targetContext = await context.ForConfiguredProjectAsync(configuredProject);
+
+                if (targetContext == null)
+                {
+                    continue;
+                }
+
+                foreach ((_, AssetsFileTargetLibrary library) in target.LibraryByName)
+                {
+                    if (context.CancellationToken.IsCancellationRequested)
+                    {
+                        // Search was cancelled
+                        return;
+                    }
+
+                    if (targetContext.IsMatch(library.Name))
+                    {
+                        targetContext.SubmitResult(CreateLibraryItem(library));
+                    }
+
+                    SearchAssemblies(library, library.CompileTimeAssemblies, PackageAssemblyGroupType.CompileTime);
+                    SearchAssemblies(library, library.FrameworkAssemblies, PackageAssemblyGroupType.Framework);
+                    SearchContentFiles(library);
+                }
+
+                SearchLogMessages();
+
+                continue;
```

Comment:
This isn't an optimisation, it's a style thing. Roslyn does it too. The idea is that local functions are towards the end and you put a `return` (in this case, continue) to demarcate the boundary.

The local functions below are not delegates. In fact the compiler can generate better code than it would for delegates, using a value type for enclosed state where possible.